e2e/func-e_use_test.go

```diff
@@ -65,3 +68,100 @@ func TestFuncEUse_UnknownVersion(t *testing.T) {
+	// The initial version.
+	baseVersion := version.Version(base)
+	// The upgraded version.
+	upgradedVersion := version.Version(upgraded)
+
+	homeDir := t.TempDir()
+
+	t.Run("install last known", func(t *testing.T) {
+		stdout, stderr, err := funcEExec("--home-dir", homeDir, "use", string(version.LastKnownEnvoy))
+
+		require.NoError(t, err)
+		require.Regexp(t, `^downloading https:.*tar.*z\r?\n$`, stdout)
+		require.Empty(t, stderr)
+
+		// The binary was installed.
+		envoyBin := filepath.Join(homeDir, "versions", string(version.LastKnownEnvoy), "bin", "envoy"+moreos.Exe)
+		require.FileExists(t, envoyBin)
+
+		// The current version was written.
+		f, err := os.ReadFile(filepath.Join(homeDir, "version"))
+		require.NoError(t, err)
+		require.Equal(t, version.LastKnownEnvoy, version.Version(f))
+	})
+
+	t.Run(fmt.Sprintf("install %s as base version", base), func(t *testing.T) {
+		stdout, stderr, err := funcEExec("--home-dir", homeDir, "use", string(baseVersion))
+
+		require.NoError(t, err)
+		require.Regexp(t, `^downloading https:.*tar.*z\r?\n$`, stdout)
+		require.Empty(t, stderr)
+
+		// The binary was installed.
+		envoyBin := filepath.Join(homeDir, "versions", string(baseVersion), "bin", "envoy"+moreos.Exe)
+		require.FileExists(t, envoyBin)
+
+		// The base version was written.
+		f, err := os.ReadFile(filepath.Join(homeDir, "version"))
+		require.NoError(t, err)
+		require.Equal(t, baseVersion, version.Version(f))
+	})
+
+	t.Run(fmt.Sprintf("install %s as upgraded version", upgraded), func(t *testing.T) {
+		stdout, stderr, err := funcEExec("--home-dir", homeDir, "use", string(minorVersion))
+
+		require.NoError(t, err)
+		require.Regexp(t, `^downloading https:.*tar.*z\r?\n$`, stdout)
+		require.Empty(t, stderr)
+
+		// The binary was installed.
+		envoyBin := filepath.Join(homeDir, "versions", string(upgradedVersion), "bin", "envoy"+moreos.Exe)
+		require.FileExists(t, envoyBin)
+
+		// The upgraded version was written.
+		f, err := os.ReadFile(filepath.Join(homeDir, "version"))
+		require.NoError(t, err)
+		require.Equal(t, minorVersion, version.Version(f))
+	})
+
+	t.Run("use upgraded version after downloaded", func(t *testing.T) {
+		stdout, stderr, err := funcEExec("--home-dir", homeDir, "use", string(minorVersion))
+		require.NoError(t, err)
+		require.Equal(t, moreos.Sprintf("%s is already downloaded\n", upgradedVersion), stdout)
+		require.Empty(t, stderr)
+	})
+
+	t.Run("which upgraded version", func(t *testing.T) {
+		stdout, stderr, err := funcEExec("--home-dir", homeDir, "which")
+		relativeEnvoyBin := filepath.Join("versions", string(upgradedVersion), "bin", "envoy"+moreos.Exe)
+		require.Contains(t, stdout, moreos.Sprintf("%s\n", relativeEnvoyBin))
+		require.Empty(t, stderr)
+		require.NoError(t, err)
+	})
+}
+
+// getVersionsRange returns the latest patch of a minor version and the one before it.
+func getVersionsRange(stdout, minor string) (max, min string) {
```

Comment:
nit: strictly speaking this is not "min", right? 😄

Reply:
Yeah, I'm changing this to `0` to make it more "versatile", but still set the purpose right.

Reply:
Updated.